### DISCUSSION TITLE
[Week8 hw2]如何不重複觸發 scroll 事件

### DISCUSSION THREAD

**Discussion by EshauHuang** [Lidemy/forum#145] [程式導師實驗計畫] [answered]
各位好，
在處理前幾週串 Twitch 的作業時，想做出滾動卷軸來加載 20 部影片的功能，目前是可以成功加載影片，但會發生一次跳出多部影片的狀況。

附上程式碼：

```
window.addEventListener('scroll', (e) => {
  const streamArea = document.querySelector('.stream-block') // 影片資訊會丟掉這裡
  const lastChild = document.querySelector('.stream-block').lastChild // 最後一部影片
  lastChildOffsetBottom = document.body.clientHeight - lastChild.offsetTop
  console.groupCollapsed('數值')
  console.log('最後一個元素頂端到底部的距離: ', lastChildOffsetBottom)
  console.log('可視範圍的距離(視窗的長度): ', window.innerHeight)
  console.log('不可視範圍距離(拉條拉到最底的長度): ', window.pageYOffset)
  console.log('body 的距離: ', document.body.clientHeight)
  console.groupEnd()
  if 
    (
      window.innerHeight + window.pageYOffset >= 
        (document.body.clientHeight - lastChildOffsetBottom)
    ) {
      const gameName = document.querySelector('.section__game-name').innerText
      getMoreStreams(gameName)
  }
})

var getMoreStreams = (gameName) => {
  offset += 20
  console.log('offset', offset)
  getStreams(gameName, appendStreams, offset)
}
```
解釋一下 `lastChildOffsetBottom` 是因為我不想碰到底部才加載影片，刪掉的話就代表到底部再加載影片，`getStreams` 為新增影片的方式，這邊應該沒有什麼問題，用按鈕來執行時沒有特別的狀況。個人認為主要是「碰到觸發事件的點」時，資料還未跑完，因此 body 的長度未能變長。

這邊觀察到 offset 的值會連加很多次，如果慢慢滑的話 offset 的值就會加的比較少。

```
offset: 20
.
.
.
offset: 140
數值
最後一個元素到底部的距離:  234
可視範圍的距離(視窗的長度):  319
不可視範圍距離(拉條拉到最底):  1371
body 的距離:  1886
offset: 160
數值
最後一個元素到底部的距離:  234
可視範圍的距離(視窗的長度):  319
不可視範圍距離(拉條拉到最底):  1372
body 的距離:  8420
```

目前有兩個想法：
 * 觸發事件時，不等資料跑完就先把區塊顯示出來
 * 用延遲的方式，但考慮到如果加載比較慢的話，延遲時間是否要拉長？因此不太想用此方法

想請問有沒有更好的方式來處理這種「觸發事件跑得比資料還快」的問題，或是我用的這個方法並不是做這個功能的好方法，感謝大家！

附上完整程式碼：[HTML](https://pastebin.com/sSUvpq6W) [Javascript](https://pastebin.com/shW7ShrN)

**Comment by allenliaooooooo** [comment, marked as answer]
滑動的時候，應該每個 px 都會觸發 scroll 事件
但我們的資料沒有這麼快回來，在等待資料回來並 append 到 DOM 上之前的每個 px 的滑動都會再發一次請求
scroll 事件不管怎樣都會觸發的，也不用限制他，只能在拿資料前先做一些判斷
我的作法是設一個變數 `isLoading`
在要拿資料之前判斷 `isLoading`，`false` 才可以拿資料

**Reply by aszx87410** [reply]
選你正解

**Reply by EshauHuang** [reply]
非常感謝！馬上來試試～

**Comment by EshauHuang** [comment]
感謝Rock大，這邊已經成功使用該方法解決問題了！

一開始思路是把「觸發事件的點」當成 isLoading 的斷點，也就是在這時將 isLoading 改為 true，但發現此方法 offset 還是會不斷增加，最後想到可以放在 getMoreStreams 內，並寫一個判斷式只要 isLoading 為 true 就跳出

```
var getMoreStreams = (gameName) => {
  if (isLoading) return
  offset += 20
  getStreams(gameName, appendStreams, offset)
  isLoading = true
}
```
另外我在 JS 剛載入時先宣告 isLoading 為 false，並在 appendStreams，也就是加入實況的元素後再把 isLoading 改為 false，讓這個事件可以繼續觸發。
